morango/models/core.py

```diff
@@ -546,32 +550,72 @@ def update_fsics(cls, fsics, sync_filter):
+        side still using a "min" (though after it has completed a sync, it updates its counters
+        such that the min and max should then be equivalent).
+
+        One potential issue remains, but it is an edge case that can be worked around:
+        - We now take the maxes across the filter partitions and use those as the producer FSICs.
+        - When the receiver finishes integrating the received data, it updates its counters to match.
+        - If the sender had actually done a sync with just a subset of those filters in the past, it
+          might not actually have received everything available for the other filters, and hence the
+          receiver may not be correct in assuming it now has everything up to the levels of the
+          producer's FSICs (as it does by taking the "max" across the filter partition FSICs).
+        There are two ways to avoid this:
+        - Don't sync with differing subsets of the same partitions across multiple syncs. For
+          example, if you do syncs with filters "AB" and "AC", don't also do syncs with filters
+          "AC" and "AD". This is the approach that makes this work in Kolibri, for now.
+        - OR: Don't do syncs with more than one filter partition at a time. Do each one in sequence.
+          For example, rather than pushing "AB" and "AC" in a single transfer session, do one pull
+          for AB and then another one for AC. This has the disadvantage of a bit of extra overhead,
+          but would likely be the most robust option, and the easiest to enforce and reason about.
+        """
 
-        # create string of prefixes to place into sql statement
-        condition = " UNION ".join(
-            ["SELECT CAST('{}' as TEXT) AS a".format(prefix) for prefix in filters]
-        )
+        queryset = cls.objects.all()
 
-        filter_max_calculation = """
-        SELECT PMC.instance, MIN(PMC.counter)
-        FROM
-            (
-            SELECT dmc.instance_id as instance, MAX(dmc.counter) as counter, filter as filter_partition
-            FROM {dmc_table} as dmc, (SELECT T.a as filter FROM ({filter_list}) as T) as foo
-            WHERE filter LIKE dmc.partition || '%'
-            GROUP BY instance, filter_partition
-            ) as PMC
-        GROUP BY PMC.instance
-        HAVING {count} = COUNT(PMC.filter_partition)
-        """.format(
-            dmc_table=cls._meta.db_table, filter_list=condition, count=len(filters)
-        )
+        per_filter_max = []
 
-        with connection.cursor() as cursor:
-            cursor.execute(filter_max_calculation)
-            # try to get hex value because postgres returns values as uuid
-            return {getattr(tup[0], "hex", tup[0]): tup[1] for tup in cursor.fetchall()}
+        for filt in filters:
+            qs = queryset.annotate(filt=Value(filt, output_field=models.CharField()))
+            qs = qs.filter(filt__startswith=F("partition"))
```

Comment:
For the above 2 lines, this seems a little bit of Django hand-holding to get the query we want. It doesn't make the intended SQL very clear at first glance, but I don't have a good solution as I've struggled with this before. I would recommend just adding a comment above summarizing the intended resulting SQL.

Reply:
Also worth noting Django adds some nested string replacements to the clause which seem unnecessary to me:
```sql
SELECT "morango_databasemaxcounter"."instance_id", MAX("morango_databasemaxcounter"."counter") AS "maxval" 
FROM "morango_databasemaxcounter" 
WHERE %s LIKE REPLACE(REPLACE(REPLACE(("morango_databasemaxcounter"."partition"), \'\\\', \'\\\\\'), \'%%\', \'\\%%\'), \'_\', \'\\_\') || \'%%\' ESCAPE \'\\\' GROUP BY "morango_databasemaxcounter"."instance_id"
```

Reply:
It looked like `HAVING {count} = COUNT(PMC.filter_partition)` enforced that all filters matched, since `{count}` was the length of the filters. So essentially what you're doing is `HAVING COUNT(PMC.filter_partition) > 0`, where at least one filter matches.

> so it would be nice to optimize it again once we've verified this produces the desired outcomes.

Sounds like maybe we should make this a beta release then?

This will get us the simplified query that we want:
```diff
diff --git a/morango/models/core.py b/morango/models/core.py
index 520b23c..abfa8e3 100644
--- a/morango/models/core.py
+++ b/morango/models/core.py
@@ -12,11 +12,15 @@ from django.db import transaction
 from django.db.models import F
 from django.db.models import Func
 from django.db.models import Max
+from django.db.models import Q
 from django.db.models import TextField
 from django.db.models import Value
 from django.db.models.deletion import Collector
+from django.db.models.expressions import CombinedExpression
 from django.db.models.fields.related import ForeignKey
 from django.db.models.functions import Cast
+from django.db.models.functions import Concat
+from django.db.models.sql.where import WhereNode
 from django.utils import six
 from django.utils import timezone
 
@@ -514,6 +518,16 @@ class AbstractCounter(models.Model):
         abstract = True
 
 
+class ValueStartsWithField(CombinedExpression):
+    def __init__(self, value, field):
+        super(ValueStartsWithField, self).__init__(
+            Value(value, output_field=models.CharField()),
+            "LIKE",
+            CombinedExpression(F(field), '||', Value("%", output_field=models.CharField())),
+            output_field=models.BooleanField()
+        )
+
+
 class DatabaseMaxCounter(AbstractCounter):
     """
     ``DatabaseMaxCounter`` is used to keep track of what data this database already has across all
@@ -593,8 +607,8 @@ class DatabaseMaxCounter(AbstractCounter):
         per_filter_max = []
 
         for filt in filters:
-            qs = queryset.annotate(filt=Value(filt, output_field=models.CharField()))
-            qs = qs.filter(filt__startswith=F("partition"))
+            qs = queryset.annotate(filter_matches=ValueStartsWithField(filt, "partition"))
+            qs = qs.filter(filter_matches=True)
             maxes = qs.values("instance_id").annotate(maxval=Max("counter"))
             per_filter_max.append({dmc["instance_id"]: dmc["maxval"] for dmc in maxes})
 ```

Reply:
That looks like a great tweak -- and yeah, a beta sounds good for now (don't see any blockers for doing one on Morango, but don't think we ever have -- I'm assuming Github Actions won't care for building it, and then the name will be parsed by PyPi, so it should just be magic. Works for me -- thanks!